src/lib/tls/tls13_pqc/kex_to_kem_adapter.cpp

```diff
@@ -0,0 +1,266 @@
+/**
+ * Adapter that allows using a KEX key as a KEM, using an ephemeral
+ * key in the KEM encapsulation.
+ *
+ * (C) 2023 Jack Lloyd
+ *     2023 Fabian Albert, René Meusel - Rohde & Schwarz Cybersecurity
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ */
+
+#include <botan/internal/kex_to_kem_adapter.h>
+
+#include <botan/internal/fmt.h>
+#include <botan/internal/pk_ops_impl.h>
+#include <botan/internal/stl_util.h>
+
+#include <variant>
+
+#if defined(BOTAN_HAS_DIFFIE_HELLMAN)
+   #include <botan/dh.h>
+   #include <botan/dl_group.h>
+#endif
+
+#if defined(BOTAN_HAS_ECDH)
+   #include <botan/ecdh.h>
+#endif
+
+#if defined(BOTAN_HAS_CURVE_25519)
+   #include <botan/curve25519.h>
+#endif
+
+namespace Botan::TLS {
+
+namespace {
+
+/**
+ * Up-cast a key agreement public key to its concrete algorithm type
+ * to perform low-level operations that are not covered by a generic
+ * interface in Public_Key.
+ */
+std::variant<std::reference_wrapper<const DH_PublicKey>,
+             std::reference_wrapper<const ECDH_PublicKey>,
+             std::reference_wrapper<const Curve25519_PublicKey>>
+as_kex_public_key(const Public_Key& kex_public_key) {
+   BOTAN_ASSERT_NOMSG(kex_public_key.supports_operation(PublicKeyOperation::KeyAgreement));
+
+   try {
+#if defined(BOTAN_HAS_ECDH)
+      if(kex_public_key.algo_name() == "ECDH") {
+         return dynamic_cast<const ECDH_PublicKey&>(kex_public_key);
+      }
+#endif
+
+#if defined(BOTAN_HAS_DIFFIE_HELLMAN)
+      if(kex_public_key.algo_name() == "DH") {
+         return dynamic_cast<const DH_PublicKey&>(kex_public_key);
+      }
+#endif
+
+#if defined(BOTAN_HAS_CURVE_25519)
+      if(kex_public_key.algo_name() == "Curve25519") {
+         return dynamic_cast<const Curve25519_PublicKey&>(kex_public_key);
+      }
+#endif
+   } catch(const std::bad_cast&) {
+      throw Invalid_Argument(fmt("public key claimed to be '{}' but down-cast failed", kex_public_key.algo_name()));
+   }
+
+   throw Not_Implemented(fmt("Cannot use unknown key agreement public key of type '{}' in the hybrid KEM key",
+                             kex_public_key.algo_name()));
+}
+
+/**
+ * This helper converts a key agreement public key into its raw public
+ * value. In contrast to the private key (cf. PK_Key_AgreementKey) there
+ * is no generic interface class to do this.
```

Comment:
Oof. This PR is definitely pointing to some areas where the API needs to be improved.

(What you're doing in this PR for this seems fine in the short term, and no need to complicate it with other machinery to make it more generic - but this should be fixed.)

Reply:
I created a ticket to collect TODOs: https://github.com/randombit/botan/issues/3706
(writeup is still to be completed)